server/jdkhttp-server/src/main/scala/sttp/tapir/server/jdkhttp/internal/ParsedMultiPart.scala

```diff
@@ -33,59 +51,153 @@ case class ParsedMultiPart(headers: Map[String, Seq[String]], body: Array[Byte])
+      }
+    }
+
+    private def lookForEndOrNewLine(): Unit = {
+      if (circularBuffer.endsWith(CRLF)) changeState(ParsePartHeaders)
+      else if (circularBuffer.endsWith(Array('-', '-'))) done = true
+    }
+
+    private def parseInitialBoundary(currentByte: Int): Unit = {
+      val foundBoundary = lookForBoundary(currentByte, boundaryBytes)
+      if (foundBoundary) {
+        changeState(LookForEndOrNewLine)
+      }
+    }
+
+    private def parsePartHeaders(): Unit = {
+      if (circularBuffer.endsWith(CRLF)) {
+        circularBuffer.getString match {
+          case "--\r\n"                          => done = true
+          case headerLine if !headerLine.isBlank => addHeader(headerLine)
+          case _                                 => changeState(ParsePartBody)
+        }
+        circularBuffer.reset()
+      } else if (circularBuffer.isFull) {
+        throw new RuntimeException(
+          s"Reached max size of $bufferSize bytes before reaching newline when parsing header."
         )
-      case None => changeState(AfterBoundary)
+      }
     }
-  }
 
-  def parseMultipartBody(httpExchange: HttpExchange, boundary: String): Seq[ParsedMultiPart] = {
-    val reader = new BufferedReader(new InputStreamReader(httpExchange.getRequestBody))
-    val initialParseState: ParseData = ParseData(empty, List.empty, Default)
-    Iterator
-      .continually(reader.readLine())
-      .takeWhile(_ != null)
-      .foldLeft(initialParseState) { case (state, line) =>
-        state.parseState match {
-          case Default if line.startsWith(boundary)           => state.changeState(AfterBoundary)
-          case Default                                        => state
-          case AfterBoundary if line.trim.isEmpty             => state.changeState(AfterHeaderSpace)
-          case AfterBoundary                                  => state.addHeader(line)
-          case AfterHeaderSpace if !line.startsWith(boundary) => state.addBody(line.getBytes())
-          case AfterHeaderSpace                               => state.completePart()
+    private def parsePartBody(currentByte: Int): Unit = {
+      bodySize += 1
+      val foundBoundary = lookForBoundary(currentByte, prefixedBoundaryBytes)
+      convertStreamToFileIfThresholdMet()
+      if (foundBoundary) {
+        val bytes = circularBuffer.getBytes
+        val bytesToWrite = bytes.slice(0, bytes.length - prefixedBoundaryBytes.length)
+        bytesToWrite.foreach(byte => stream.underlying.write(byte.toInt))
+        val bodyInputStream = stream match {
+          case FileStream(tempFile, stream) => stream.close(); new FileInputStream(tempFile)
+          case ByteStream(stream)           => new ByteArrayInputStream(stream.toByteArray)
         }
+        completePart(bodyInputStream)
+        stream = ByteStream()
+        bodySize = 0
+      } else if (numMatchedBoundaryChars == 0) {
+        circularBuffer.getBytes.foreach(byte => stream.underlying.write(byte.toInt))
```

Comment:
hm are you really using `CircularBuffer` as a circular buffer, not only as a buffer? It seems you are buffering bytes only when there's some agreement on the boundary bytes, but as soon as there's a mismatch the bytes are copied to the sink, and the buffer is reset (by setting its `currentIndex` to -1). So it never wraps?

Reply:
It seems you are correct :smile: 
I went through a few iterations, so i don't think we need the wrapping anymore. I'll remove it and rewrite with a normal arraybuffer.